## Data Analysis and Visualisation

In [22]:
import pandas as pd
import seaborn as sns
import matplotlib as plt
import nltk
from IPython.display import display

#### I got two datasets with me which are basically the same thing. <br> I got one which is the original raw dataset downloaded from kaggle <br> the other however is a cleaned variant of the original which i made to make future analysis and preprocessing easier

In [2]:
raw = pd.read_csv("data/mental_disorders_reddit.csv")
cleaned = pd.read_csv("data/cleaned_reddit.csv")

/tmp/ipykernel_4723/3528956614.py:1: DtypeWarning: Columns (2,3,5,6,7,9,10) have mixed types. Specify dtype option on import or set low_memory=False.
  raw = pd.read_csv("data/mental_disorders_reddit.csv")


In [ ]:
# This command basically tells us the number of (rows, columns) in the datasets
# We can see here that the original had 11 rows but my cleaned version has 2 only
# Why is that? Because in this project we will be working on those two only which you will understand why later.
print(raw.shape)
print(cleaned.shape)

(691498, 11)
(562437, 2)


In [ ]:
# This code shows us the first and last 5 rows of each row.
display(raw.head())
display(raw.tail())

,title,selftext,created_utc,over_18,subreddit,Unnamed: 5,Unnamed: 6,Unnamed: 7,Unnamed: 8,Unnamed: 9,Unnamed: 10
0,Life is so pointless without others,Does anyone else think the most important part...,1650356960,False,BPD,NaN,NaN,NaN,NaN,NaN,NaN
1,Cold rage?,Hello fellow friends 😄\n\nI'm on the BPD spect...,1650356660,False,BPD,NaN,NaN,NaN,NaN,NaN,NaN
2,I don’t know who I am,My [F20] bf [M20] told me today (after I said ...,1650355379,False,BPD,NaN,NaN,NaN,NaN,NaN,NaN
3,HELP! Opinions! Advice!,"Okay, I’m about to open up about many things I...",1650353430,False,BPD,NaN,NaN,NaN,NaN,NaN,NaN
4,help,[removed],1650350907,False,BPD,NaN,NaN,NaN,NaN,NaN,NaN


,title,selftext,created_utc,over_18,subreddit,Unnamed: 5,Unnamed: 6,Unnamed: 7,Unnamed: 8,Unnamed: 9,Unnamed: 10
691493,Is it psychosis?,"Good afternoon every one, I'm here to please f...",1589562851,False,mentalillness,NaN,NaN,NaN,NaN,NaN,NaN
691494,Looking For Music and Gaming Friends With Ment...,Hi I'm Lake (31M)! Every now and again I put ...,1589561920,False,mentalillness,NaN,NaN,NaN,NaN,NaN,NaN
691495,Anyone feel more “aware”? With mental illness,NaN,1589556993,False,mentalillness,NaN,NaN,NaN,NaN,NaN,NaN
691496,Something was said to me I that hurt worse tha...,I made a mistake in an online Facebook comment...,1589556913,False,mentalillness,NaN,NaN,NaN,NaN,NaN,NaN
691497,***Help! ***,**seeking advice***\nI’ve never been to a ther...,1589552838,False,mentalillness,NaN,NaN,NaN,NaN,NaN,NaN


In [25]:
display(cleaned.head())
display(cleaned.tail())

,text,label
0,Does anyone else think the most important part...,BPD
1,Hello fellow friends 😄\n\nI'm on the BPD spect...,BPD
2,My [F20] bf [M20] told me today (after I said ...,BPD
3,"Okay, I’m about to open up about many things I...",BPD
4,"Without going into detail, this diagnosis expl...",BPD


,text,label
562432,Heyo all yesterday my life was changed forever...,mentalillness
562433,"Good afternoon every one, I'm here to please f...",mentalillness
562434,Hi I'm Lake (31M)! Every now and again I put ...,mentalillness
562435,I made a mistake in an online Facebook comment...,mentalillness
562436,**seeking advice***\nI’ve never been to a ther...,mentalillness


#### So what happened exactly is we got rid of the strange columns named unnamed:5-10 <br> which can be caused by several things including how pandas reads CSV files sometimes and how it handles different formats.

In [35]:
# Using the display command shows us that the dataset originally is misaligned or broken...
# This is common due to how it was collected from reddit
# To fix this i dropped all columns except for selftext and subreddit columns and then proceeded to rename both to text and lebel
display(raw.describe())
print()
display(cleaned.describe())

,Unnamed: 8
count,1.000000e+00
mean,1.626848e+09
std,NaN
min,1.626848e+09
25%,1.626848e+09
50%,1.626848e+09
75%,1.626848e+09
max,1.626848e+09


,text,label
count,562437,562437
unique,555716,6
top,I’m so alone just existing for the sake of it ...,BPD
freq,122,204575


##### Above we can see how in the cleaned dataset we have only 562437 entries with 555716 being unique <br> This is because i removed all NaN or the removed subreddits from the dataset as they are considered noise <br> still there is more cleaning to be done such as removing duplicate texts, short texts that carry no value, repeated characters...etc <br> This is not covered in this notebook however but in the code file named data_preprocessing.

In [41]:
display("Data types of the original dataset",raw.dtypes)
print()
display("Data types of the cleaned dataset", cleaned.dtypes)

'Data types of the original dataset'

title           object
selftext        object
created_utc     object
over_18         object
subreddit       object
Unnamed: 5      object
Unnamed: 6      object
Unnamed: 7      object
Unnamed: 8     float64
Unnamed: 9      object
Unnamed: 10     object
dtype: object

'Data types of the cleaned dataset'

text     object
label    object
dtype: object

In [42]:
display("number of texts for each label in original dataset", raw['subreddit'].value_counts())

'number of texts for each label in original dataset'

subreddit
BPD                                                                                                                   241110
Anxiety                                                                                                               173990
depression                                                                                                            156972
bipolar                                                                                                                51109
mentalillness                                                                                                          42911
schizophrenia                                                                                                          25365
FALSE                                                                                                                      2
 and I told her that I’d go once I move out west because I didn’t want to have to deal with looking for a provider 

In [43]:
display("number of texts for each label in the cleaned dataset", cleaned['label'].value_counts())

'number of texts for each label in the cleaned dataset'

label
BPD              204575
Anxiety          160133
depression       120954
bipolar           35333
mentalillness     29920
schizophrenia     11522
Name: count, dtype: int64

##### Do note how for the original dataset we had some strange labels such as FALSE, the long text and that other one that is a number <br> This tells you bad it is to use datasets collected from Reddit, however during the time this project was being made, this was probably the best free mental health related dataset relevant to this project. <br> If you find something better to use, ofcourse feel free to do so :)